### PR TITLE
feat: new method for choice fail

### LINKED ISSUE #2621: Allow customizing fail message for invalid choice in `click.types.Choice`
Move the fail art of `convert()` in the `Choice` class to a new method `get_invalid_choice_fail_message(self)` so in my subclass of `click.Choice` I can customize the way it shows the failing value for the choice

Some CLI option choices are multi-part meaning they are separated by some value, like a comma separated list for example. Maybe each section of this value represents a different part of a schema and it is nice to show a custom message about the specific part.

Contrived version:
Let says my choices are `["maine::goat", "maine::themepark"]` and I give it "maine::toast"`, it'd be nice to say the `maine` part was correct and the `toast` part was not.

### PR DESCRIPTION
**Pull request by antazoey**
Refactors so the failing happens in a new public method.

- fixes #2621
- 
Ensure each step in CONTRIBUTING.rst is complete by adding an "x" to
each box below.

If only docs were changed, these aren't relevant and can be removed.
-->

Checklist:

- [x] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [x] Add or update relevant docs, in the docs folder and in code.
- [x] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [x] Add `.. versionchanged::` entries in any relevant code docs.
- [x] Run `pre-commit` hooks and fix any issues.
- [x] Run `pytest` and `tox`, no tests failed.

### CODE REVIEW

File: src/click/types.py [hunk 1]
```diff
@@ -297,16 +297,15 @@ def convert(
         if normed_value in normed_choices:
             return normed_choices[normed_value]
 
+        self.fail(self.get_missing_message(value), param, ctx)
```
Comment (AndreasBackx) [inline src/click/types.py:300]: This does not call the new function you created.

File: src/click/types.py [hunk 2]
```diff
@@ -297,16 +297,15 @@ def convert(
...
+
+    def get_invalid_choice_fail_message(self, value: t.Any) -> str:
```
Comment (AndreasBackx) [inline src/click/types.py:302]:
```suggestion
    def get_invalid_choice_message(self, value: t.Any) -> str:
```

Let's shorten this a bit?

**Review by AndreasBackx — CHANGES_REQUESTED** [review]
This seems like a fine choice. Could you follow through with the rest of the checklist like documentation and tests?

File: src/click/types.py (hunk 1 above)
Comment (antazoey) [reply src/click/types.py]: Resolved!

**Comment by antazoey** [comment]
I did my best to complete the checkboxes. Notes:

* I wasn't sure where to put the test as there were no other Choice-based tests in types module however the places testing choice were more integraion-y and less about the helper messages; so the test I added stands out like a lone wolf.
* Nothing is really "changing" so I opted for now `.. versionchanged::
* I didn't see any `docs/` relevant to making custom choice types so I didn't any the top-level use-case there.

**Comment by AndreasBackx** [comment]
@antazoey thanks for getting back so quickly and my apologies for taking another week myself. I've wrapped it up and merged it, thank you for this fix!

**Pull request merged**